1단계 - 문자열 덧셈 계산기

devFancy opened the pull request:
안녕하세요. 리뷰어님!
이번 미션에서는 요구사항 충족에 집중하며 간결하고 직관적인 코드를 작성하고자 했습니다. 
아래는 이번 구현의 주요 내용과 의도입니다.

---

## 구현 내용

### 클래스 설계
단일 책임 원칙(Single Responsibility Principle)을 고려했지만, 현재 요구사항에서는 지나치게 세분화하는 것보다 **이해하기 쉽고 관리하기 용이한 구조가 더 적합**하다고 판단했습니다.
따라서 다음과 같이 두 개의 클래스로 나누었습니다.
- `StringCalculator`: 문자열 덧셈 계산을 담당하는 핵심 클래스입니다.
- `StringSplitter`: 문자열을 특정 구분자로 분리하는 역할을 담당합니다.

### 요구사항
모든 요구사항을 충족하였으며, 세부 요구사항은 `1단계_요구사항.md` 파일에 정리했습니다.

---

구현한 코드에서 이해가 어려운 부분이나 부족한 점이 있다면 언제든 편하게 말씀해주세요!
리뷰 잘 부탁드립니다 ! 😊
감사합니다.

devFancy committed:
feat(racingcar): Car 및 테스트코드 작성

devFancy committed:
feat(racingcar): Junit 으로 테스트코드 작성

devFancy committed:
docs: 요구사항 및 힌트에 대한 문서화 정리

devFancy committed:
feat: 문자열 덧셈 계산기에 대한 클래스 생성

devFancy committed:
feat: 구분자를 위한 클래스 생성

devFancy committed:
feat: 예외에 대한 커스텀 처리를 위해 예외 클래스 생성

devFancy committed:
chore: 의도를 분명히 하기 위해 클래스명 수정

devFancy committed:
chore: editorconfig 파일 추가

- 마지막 개행 추가

devFancy committed:
chore: 마지막 개행 추가

mj950425 commented on src/main/java/calculator/StringAdditionCalculator.java:
```diff
@@ -0,0 +1,47 @@
+
+public class StringAdditionCalculator {
+
+    private String text;
```
클래스 변수에도 final을 붙여주는것은 어떨까요?

devFancy replied:
해당 부분 수정했습니다!

mj950425 commented on src/main/java/calculator/StringAdditionCalculator.java:
```diff
@@ -0,0 +1,47 @@
+            return 0;
+        }
+
+        String[] numbers = StringSplitter.split(text);
```
현재 문자열 계산기가 과도한 책임을 가지고 있는 것처럼 보입니다. 

객체지향적인 코드를 위해, 문자열 계산기가 처리하는 숫자를 객체로 추상화해보는 것은 어떨까요?

그렇다면 문자열 계산기가 가지고 있는 책임을 나눠가질 수 있을 것 같습니다.

devFancy replied:
넵! 그래서 숫자를 검증하고 정수를 변환하는 역할을 `PositiveNumber` 클래스를 생성해서 책임을 위임했습니다!

mj950425 replied:
```
String[] numbers = StringSplitter.split(text);
```

배열 대신 콜랙션을 사용해보는것은 어떨까요? 

아니면 따로 배열을 사용하신 이유가 있을까요?

devFancy replied:
현재 코드에서 확장성을 고려하지 못한 것 같습니다.
컬렉션 사용하여 반영했습니다!

mj950425 commented on .editorconfig:
```diff
@@ -0,0 +1,11 @@
+root = true
```
👍

mj950425 reviewed (changes requested):
고생 많으셨습니다 준용님 👍

> 단일 책임 원칙(Single Responsibility Principle)을 고려했지만, 현재 요구사항에서는 지나치게 세분화하는 것보다 이해하기 쉽고 관리하기 용이한 구조가 더 적합하다고 판단했습니다.

현재 요구사항이 복잡한것은 아니지만, 의식적으로 클린 코드를 연습하는 과정의 미션인만큼 말씀해주신 단일책임원칙에 맞게 코드를 작성해보는것도 좋아보입니다.

관련해서 짧게 코멘트 남겼습니다. 확인 부탁드려요 💪

mj950425 commented on src/test/java/calculator/StringAdditionCalculatorTest.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @DisplayName("만약 빈 문자열 또는 null을 입력할 경우, 0을 반환한다.")
+    @ParameterizedTest
+    @NullAndEmptySource
```
👍

mj950425 commented on src/main/java/calculator/StringSplitter.java:
```diff
@@ -0,0 +1,24 @@
+package calculator;
+
+public class StringSplitter {
```
생성자의 접근 제한자를 private으로 설정하면 유틸리티 클래스가 인스턴스화되는 것을 방지할 수 있습니다. 

```suggestion
    private StringSplitter() {
    }
```

devFancy replied:
유틸리티 클래스를 명확히 표현하기 위해 접근 제한자 부분까지 신경써야 한다는 점을 잊고 있었네요 ! 꼼꼼한 피드백 감사드려요!

mj950425 replied:
~~반영이 안되어있네요!~~

devFancy replied:
아래와 같이 생성자를 private로 반영했어요 !
```java
private StringSplitter() {
        throw new UnsupportedOperationException("StringSplitter is a utility class and cannot be instantiated.");
    }
```

mj950425 replied:
아하 제가 코드 확인을 못한 것 같습니다. 잘 반영해주셨네요 👍

devFancy committed:
refactor: 숫자를 검증하고 변환하는 역할을 분리

devFancy committed:
chore: 중복된 메서드 제거 및 일부 변수 수정

devFancy committed:
Merge branch 'devfancy' into step1

devFancy committed:
refactor: 생성자의 접근 제한자를 private로 설정

- private 설정함으로서 유틸리티 클래스가 인스턴스화 되는 것을 방지
- 유틸리티 클래스임을 명확히 표현하기 위해 생성자 호출시 예외 던지도록 변경

devFancy committed:
Merge branch 'step1' of https://github.com/devFancy/ddd-legacy into step1

mj950425 commented on src/main/java/calculator/StringAdditionCalculator.java:
```diff
@@ -0,0 +1,35 @@
+        String[] numbers = StringSplitter.split(text);
+        int sum = 0;
+
+        for (String number : numbers) {
```
일급 콜랙션을 사용해보는것은 어떨까요?

래퍼런스 : https://jojoldu.tistory.com/412

devFancy replied:
넵! 반영했습니다!

mj950425 commented on src/main/java/calculator/StringAdditionCalculator.java:
```diff
@@ -0,0 +1,35 @@
+package calculator;
+
+import calculator.exception.InvalidNumberFormatException;
```
사용하지않는 Import 문은 제거해도될것같아요

devFancy replied:
반영했습니다!~

mj950425 reviewed (commented):
빠른 리뷰 반영 고생하셨어요 💪

몇가지 코멘트만 추가적으로 남겨뒀는데 확인 부탁드릴게요.

devFancy committed:
refactor: 확장성을 위해 컬렉션과 Stream 을 이용

devFancy committed:
refactor: 일급 컬렉션을 위해 PositiveNumbers 클래스를 생성

- 숫자 검증 및 값을 반환은 `PositiveNumber`
- 여러 PositiveNumber 객체를 관리하고 합을 구하는 역할은 `PositiveNumbers`

devFancy committed:
docs: 참고자료 부분 추가

mj950425 commented on src/main/java/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * 숫자 검증 및 변환하는 역할
+ */
+public class PositiveNumber {
```
책임에 맞게 객체를 분리했으니, 테스트 코드도 분리할 수 있을 것 같아요.

mj950425 commented on src/main/java/calculator/exception/PositiveNumbers.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * 0 이상의 숫자들의 리스트를 관리하는 일급 컬렉션
+ */
+public class PositiveNumbers {
```
잘 만들어주셨네요 👍

만들어주신 일급 컬랙션도 테스트 코드를 만들 수 있을 것 같네요.

devFancy replied:
이 부분은 2단계 미션 진행하면서 같이 짜보도록 할게요!

mj950425 commented on src/main/java/calculator/exception/PositiveNumbers.java:
```diff
@@ -0,0 +1,26 @@
+    public PositiveNumbers(final List<String> numberToStrings) {
+        this.numbers = numberToStrings.stream()
+            .map(PositiveNumber::new)
+            .collect(Collectors.toList());
```
자바 21을 사용하고 계시네요. 

자바 16부터는 이렇게 고칠 수 있어요.

```suggestion
            .toList();
```

devFancy replied:
자바 11 이후의 문법을 제대로 몰랐는데, 미션을 진행해보면서 틈틈히 공부해봐야겠네요!
꼼꼼한 피드백 감사드려요!

devFancy replied:
2단계 미션에 반영해두겠습니다!

mj950425 reviewed (approved):
낮부터 고생하셨습니다 👍👍 

마이너한 코멘트만 남겨두었고, 이번 스탭은 머지하도록 하겠습니다~

다음 미션으로 가시죠!
